### DISCUSSION TITLE
Create file if it doesn't exist, fail otherwise

### DISCUSSION THREAD

**Discussion by retog** [denoland/deno#12903] [Q&A] [answered]
In my application, I need to create a file if it doesn't exist but if the file already exist this should trigger some special handling. This can be done synchronously: 
```
writeNewFile(filename, data) {
  try {
    Deno.statSync(filename);
    // successful, file or directory must exist
    throw Error("There's alredy such a file");
  } catch (error) {
    if (error instanceof Deno.errors.NotFound) {
      // file or directory does not exist
      Deno.writeFileSync(filename, data);
    } else {
      // unexpected error, maybe permissions, pass it along
      throw error;
    }
  }
```
However, as my applications handles many requests asynchronously and writing such files is a frequent, writing the files synchronously isn't satisfactory.  Using the async `Deno.writeFile`-method would entail the risk that before the file has been written, a second invocation of the function doesn't yet find an existing file and also starts writing.

Is there a way to asynchronously write a file only if it doesn't yet exist?

**Comment by bartlomieju** [comment, marked as answer]
@retog your solution is quite racy. You can do that using `Deno.open()` API and specifying relevant options:
```js
import { writeAll } from "https://deno.land/std/streams/conversion.ts";
const file = await Deno.open(filename, { createNew: true, write: true });
await writeAll(file, content)
Deno.close(file.rid);
```
`createNew` option ensures that no file exists at specified location, otherwise it will throw an error.

See https://doc.deno.land/builtin/stable#Deno.OpenOptions for more details.

**Reply by retog** [reply]
Great, that's very helpful, thanks a lot!

I'd suggest changing the example to the following:
```
import { writeAll } from "https://deno.land/std/streams/conversion.ts";
const file = await Deno.open(filename, { createNew: true, write: true });
await writeAll(file, content)
Deno.close(file.rid);
```

**Reply by bartlomieju** [reply]
Thanks, I updated my example.
